### ISSUE TITLE
Add activities to SignalR invocations

### ISSUE THREAD

**Issue by davidfowl** [dotnet/aspnetcore#51557]
Today SignalR invocations do not have any activity associated with them. We currently piggy-back on the default ASP.NET Core activity and it's not great for invocations over the websocket. We should add an activity following the [semantic conventions for rpc](https://opentelemetry.io/docs/specs/semconv/rpc/rpc-spans/).

Open questions:
- Should we do this tracing in all directions? Client to server and server to client? 
- Do we use message headers to pass this state across the wire seamlessly? This will impact the performance of invocations.

**Comment by BrennanConroy** [comment]
Related
https://github.com/dotnet/aspnetcore/issues/18711
https://github.com/dotnet/aspnetcore/issues/40763

**Comment by VincentH-Net** [comment]
> Should we do this tracing in all directions? Client to server and server to client?

Yes please. Bidirectional tracing is useful in scenario's like IoT devices with local intelligent hubs plus cloud (a project I am working on). Or like in [this chat app example trace](https://twitter.com/vincenth_net/status/1747214763996672337), 2 app instances connected to the cloud backend with SignalR. It is helpful to see the true end2end flow and timings.

**Comment by adityamandaleeka** [comment]
Still planning to do this in 9.

**Comment by BrennanConroy** [comment]
The activity source `"Microsoft.AspNetCore.SignalR.Server"` can now be listened to on server applications.

- [x] Server activity source ("Microsoft.AspNetCore.SignalR.Server")
- [ ] Client activity source
- [x] Hub method events
- [ ] Client method events
- [ ] Distributed tracing (trace ids sent between client and server)

**Comment by adityamandaleeka** [comment]
Lots of progress on this in .NET 9. Moving to 10 for the rest.

**Comment by adityamandaleeka** [comment]
Actually, this might be worth reconsidering a bit more thoughtfully to decide which parts to prioritize. Un-milestoning for now (cc @BrennanConroy @JamesNK)

**Comment by paulomorgado** [comment]
An `Activity` for the handling server notifications is still missing.

And it would be nice to add connection ID to the spans.

**Comment by BrennanConroy** [comment]
> An `Activity` for the handling server notifications is still missing.

What do you mean by this? There is a server activity "Microsoft.AspNetCore.SignalR.Server".

Is what you're trying to ask about listed in the list of things left to do? https://github.com/dotnet/aspnetcore/issues/51557#issuecomment-2136051137

**Comment by paulomorgado** [comment]
> What do you mean by this? There is a server activity "Microsoft.AspNetCore.SignalR.Server".

I'm talking about client side: "Microsoft.AspNetCore.SignalR.Client"

> Is what you're trying to ask about listed in the list of things left to do? [#51557 (comment)](https://github.com/dotnet/aspnetcore/issues/51557#issuecomment-2136051137)

Probably included in "Client method events" and "Distributed tracing (trace ids sent between client and server)/Server to Client".

**Comment by ErisApps** [comment]
Took me quite some time to realize, as I had misinterpreted the documentation (https://learn.microsoft.com/en-us/aspnet/core/signalr/diagnostics?view=aspnetcore-10.0#net-signalr-client-activitysource) apparently despite double- and triple-checking 😅, but finally figured out due to stumbling upon this issue and looking through source that Server -> Client invocations don't generate spans just yet.

@BrennanConroy Is it still planned, as this issue was added to the .NET 11 Planning milestone, but seemingly not listed on the .NET 11 Roadmap (https://github.com/dotnet/aspnetcore/issues/64787)? And if so, targeted for .NET 11 or rather an eventual future release?
